packages/core/parcel-bundler/src/Server.js

```diff
@@ -68,15 +66,33 @@ function middleware(bundler) {
     }
 
     function sendIndex() {
-      // If the main asset is an HTML file, serve it
-      if (bundler.mainBundle.type === 'html') {
-        req.url = `/${path.basename(bundler.mainBundle.name)}`;
+      const {pathname} = url.parse(req.url);
+
+      // Try to find the index file. If not found, return a 404 error
+      const index = findIndex(pathname);
+      if (index) {
+        req.url = `/${path.basename(index)}`;
         serve(req, res, send404);
       } else {
         send404();
       }
     }
 
+    function findIndex(pathname) {
+      if (bundler.mainBundle.type === 'html') {
+        // If the bundle is HTML, send that
+        return bundler.mainBundle.name;
+      } else if (pathname === bundler.options.publicURL) {
+        // if the pathname is empty, look for files named index.html in the bundle and send back the first one. If there's nothing, return null;
+        const indexes = Array.from(bundler.mainBundle.childBundles).filter(
+          asset => path.basename(asset.name) === 'index.html'
+        );
+        return indexes[0].name;
+      }
```

Comment:
It seems like the logic here should be:

1. Is there a file that matches something like `${req.url}/index.html`, then serve that
2. If not, then is there an `index.html` at the root of the dist directory, then serve that
3. Otherwise 404

IDK if that’s kind of special-cased or not, but when you only specify one entry point, `parcel serve` always returns that file for *any* url. In my situations where I have had multiple entry points, I’ve wanted the root index file to be the catch-all and any other specific ones to be accessible directly. My use-case is having an index.html at the root running React Router and another file that lives outside of my react app that can be accessed directly.

Reply:
Hey @mrcoles, thanks for your review! If I understand you correctly you want the order of resolving to be as you stated in your list (first specified file, then index.html, then 404). 

The thing I don't understand is how that's not the case right now? If I have two entry points, index.html and one.html and I point my browser to /one.html it'll serve one.html, otherwise it'll serve index.html (at least, it should). If you specify only one entry point, this code should

Reply:
Hey @LuudJanssen it’s possible I’m misinterpreting what `bundler.options.publicURL` and `bundler.mainBundle.childBundles` are, but from looking at just this code it seems like the following edge cases could happen:

1. if there are multiple index.html entry points, it seems like this would randomly pick one of them (or more specifically whichever is the first one in `bundler.mainBundle.childBundles`—is that just how they’re ordered in the cli command?)

2. If there are *no* index.html entry points, but there are multiple entry points, then `indexes[0].name` will raise an error, because `indexes[0]` will be `undefined`. (Also for the sake of consistency, if none are found, it should probably return `null` to match the other case where `null` is returned)

Do you think it would make sense to add some additional tests for when there are multiple index.html files and when there are none?

Reply:
Oh yeah, I am misinterpreting some of this (I was misunderstanding `pathname === bundler.options.publicURL`), but it looks like there are still some edge cases to handle?